go/registry/api/api.go

```diff
@@ -908,3 +908,110 @@ type ConsensusParameters struct {
 	// related checks and operations.
 	DebugBypassStake bool `json:"debug_bypass_stake"`
 }
+
+// SanityCheck does basic sanity checking on the genesis state.
+func (g *Genesis) SanityCheck() error {
+	if !g.Parameters.KeyManagerOperator.IsValid() {
+		return fmt.Errorf("registry: sanity check failed: key manager operator's public key is invalid")
+	}
+
+	// Check entities.
+	seenEntities := make(map[signature.MapKey]bool)
+	for _, sent := range g.Entities {
+		var ent entity.Entity
+		if err := sent.Open(RegisterGenesisEntitySignatureContext, &ent); err != nil {
+			return fmt.Errorf("registry: sanity check failed: unable to open signed entity")
+		}
+
+		if !ent.ID.IsValid() {
+			return fmt.Errorf("registry: sanity check failed: entity ID %s is invalid", ent.ID.String())
+		}
+
+		for _, pk := range ent.Nodes {
+			if !pk.IsValid() {
+				return fmt.Errorf("registry: sanity check failed: entity ID %s has node with invalid ID %s", ent.ID.String(), pk.String())
+			}
+		}
+
+		if ent.RegistrationTime > uint64(time.Now().Unix()+61*60) {
+			return fmt.Errorf("registry: sanity check failed: entity ID %s registration time is more than 1h1m in the future", ent.ID.String())
+		}
+
+		seenEntities[ent.ID.ToMapKey()] = true
+	}
+
+	// Check runtimes.
+	seenRuntimes := make(map[signature.MapKey]bool)
+	for _, srt := range g.Runtimes {
+		var rt Runtime
+		if err := srt.Open(RegisterGenesisRuntimeSignatureContext, &rt); err != nil {
+			return fmt.Errorf("registry: sanity check failed: unable to open signed runtime")
+		}
+
+		if !rt.ID.IsValid() {
+			return fmt.Errorf("registry: sanity check failed: runtime ID %s is invalid", rt.ID.String())
+		}
+
+		if rt.RegistrationTime > uint64(time.Now().Unix()+61*60) {
+			return fmt.Errorf("registry: sanity check failed: runtime ID %s registration time is more than 1h1m in the future", rt.ID.String())
+		}
+
+		if !rt.KeyManager.IsValid() {
```

Comment:
Should probably also cross-check that the keymanager is actually a valid key manager runtime.